File: c++/85_Maximal_Rectangle.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>
#include <stack>

using namespace std;


class Solution {
public:
    int maximalRectangle(vector<vector<char> >& matrix) {
    	int m = matrix.size();
    	if (m == 0) return 0;
    	int n = matrix[0].size();
    	if (n == 0) return 0;
    	vector<int> dp(n, 0);
    	for (int i = 0; i < n; i++)
    		if (matrix[0][i] == '1')
    			dp[i] = 1;
    	int maxS = largestRectangleArea(dp); // use function from #84
    	for (int i = 1; i < m; i++) {
    		for (int j = 0; j < n; j++) {
    			if (matrix[i][j] == '1') // from top to bottom, iteration, reference from youTube
    				dp[j] += 1; 
    			else
    				dp[j] = 0;
    		}
    		maxS = max(maxS, largestRectangleArea(dp));
    	}
    	return maxS;
    }

    int largestRectangleArea(vector<int>& heights) {
    	if (heights.size() == 0) return 0;
    	stack<int> stk;
    	heights.push_back(-1);
    	heights.insert(heights.begin(), -1);
    	int maxS = 0;
    	for (int i = 0; i < heights.size(); i++) {
    		while (!stk.empty() and heights[i] <= heights[stk.top()]) {
    			int minH = heights[stk.top()];
    			stk.pop();
    			if (!stk.empty()) maxS = max(maxS, (i - stk.top() - 1) * minH);
    		}
    		stk.push(i);
    	}
    	heights.pop_back();
    	heights.erase(heights.begin());
    	return maxS;
    }
};


int main() {
	Solution s;
}

File: python/85_Maximal_Rectangle.py
#!/usr/bin/python
import unittest

'''
1 0 1 0 0
1 0 1 1 1
1 1 1 1 1
1 0 0 1 0
first use dp


'''

class Solution(object):
    def maximalRectangle(self, matrix):
    	m = len(matrix)
    	if m: 
    		n = len(matrix[0])
    		if n == 0: return 0
    	else:
    		return 0
    	dp = [0] * n
    	for j in range(n):
    		if matrix[0][j] == "1":
    			dp[j] = 1
    	maxS = self.largestRectangleArea(dp) # trival bug in the following function
    	dp.remove(-1)
    	for i in range(1, m):
    		for j in range(n):
    			if matrix[i][j] == "1":
    				dp[j] = dp[j] + 1
    			else: dp[j] = 0
    		maxS = max(maxS, self.largestRectangleArea(dp))
    		dp.remove(-1)
    	return maxS
    def largestRectangleArea(self, heights):
    	if heights == []: return 0
    	maxArea = heights[0]
    	stk = []
    	heights.append(-1)
    	heights = [-1] + heights # make sure stk always has a 0
        for i in range(len(heights)):
        	while stk and heights[i] <= heights[stk[-1]]:
        		minH = heights[stk[-1]]
        		stk.pop()        	
        		if stk: # when it comes to the last item, both minH and height[i] can be -1, so stk is empty
        			maxArea = max(maxArea, (i - stk[-1] - 1) * minH)
        	stk.append(i)
        heights.remove(-1)
        heights.remove(-1)
        return maxArea







class TestSolution(unittest.TestCase):
    def setUp(self):
        self.s = Solution()

    def test_method(self):
        """Such as self.assertEqual, self.assertTrue"""
        self.assertEqual(self.s.maximalRectangle(["10100","10111","11111","10010"]), 6)
        pass

def main():
    s = Solution()


if __name__ == "__main__":
    unittest.main()
